### PR TITLE
Disallowing null values in ContextSnapshot

### PR DESCRIPTION
**Pull request by chemicL**
This change takes care of guaranteeing `@NonNullApi` package annotation is in effect for `ThreadLocalAccessor` calls when dealing with `ContextSnapshot`. When a source context `Object` provided `null` for a key, the API would not have held its guarantees. Now we explicitly make sure the backing `Map` has no mappings to `null`.

### CODE REVIEW

**Comment by rstoyanchev** [comment]
A bit more context from offline conversations. 

When capturing `ThreadLocal` values, we already [prevent `null`](https://github.com/micrometer-metrics/context-propagation/blob/dba041807cf70ba8ef84dd8e250a1d040a0eea17/context-propagation/src/main/java/io/micrometer/context/DefaultContextSnapshot.java#L147) from going into a `ContextSnapshot`, and with a `ThreadLocal` it's impossible to differentiate between a `ThreadLocal` with `null` vs no `ThreadLocal` at all. Likewise the Reactor `Context` does not allow `null` values, so there is no way for `null` values to enter a `ContextSnapshot` that way either.

Nevertheless, since `ContextSnapshot` is essentially a `Map`, it is possible for some other type of context and `ContextAccessor` to add `null` values. The possibility of `null` values in the `ContextSnapshot` creates ambiguity since it's not clear what to do with that when setting `ThreadLocal` values, i.e. whether to clear it or set it to `null`. Furthermore, as mentioned above, as it is now, it's not possible even to add `null` values into a `ContextSnapshot` in many cases. So rather than trying to come with with ways to deal with `null` values`, it is better to close the loophole, and guarantee the `@NonNullApi` contract for values.

File: context-propagation/src/test/java/io/micrometer/context/DefaultContextSnapshotTests.java [hunk 1]
```diff
@@ -53,6 +53,24 @@ void should_propagate_thread_local() {
...
+        Map<String, String> sourceContext = Collections.singletonMap(key, emptyValue);
+
+        ContextSnapshot snapshot = ContextSnapshot.captureAll(this.registry, sourceContext);
+
+        try (Scope scope = snapshot.setThreadLocals()) {
+            assertThat(ObservationThreadLocalHolder.getValue()).isEqualTo(emptyValue);
+        }
+        assertThat(ObservationThreadLocalHolder.getValue()).isEqualTo(emptyValue);
+    }
+
```
Comment (rstoyanchev) [inline context-propagation/src/test/java/io/micrometer/context/DefaultContextSnapshotTests.java:73]: This test method could go further down in the overall order. Right now it's in the middle of `should_propagate_*` tests. Perhaps, it can go right after `should_filter_*` tests.

It would also be useful to have several tests around this, to cover all the ways in which a `null` value could potentially make its way in:

1. capturing from `ThreadLocal` values into a `ContextSnapshot`
2. capturing from a context into a `ContextSnapshot (covered)
3. setting all `ThreadLocal` values directly from a context
4. setting a subset of `ThreadLocal` values from a context.

For the implementation, one thought is to downcast the `ContextSnapshot` to a `Map` since we know that's what is created by default, and check for keys. Seems more transparent and easier to understand, and such an assumption in tests would be okay.

**Review by rstoyanchev — CHANGES_REQUESTED** [review]
Overall, good changes!

1. I would like to see a Javadoc updates that settles this topic more formally in writing now that we've looked at it in more detail and made concrete decisions. Starting with `ContextSnapshot`, clarify that a snapshot should not contain `null` values since those are not useful for setting `ThreadLocal` values anyway, and that all implementations should ensure that `null` values are ignored, and never passed into `ContextAccessor` or `ThreadLocalAccessor`. Similar updates could be made in `ContextAccessor`, e.g. on `readValues` to explain that any `null` values put into the `Map` will be filtered out, and likewise on `setValues` to clarify the source `Map` will never have `null` values. Make similar updates in `ThreadLocalAccessor` too in the `setValue` and `restore(String)` methods.
2. Around [this line](https://github.com/micrometer-metrics/context-propagation/blob/dba041807cf70ba8ef84dd8e250a1d040a0eea17/context-propagation/src/main/java/io/micrometer/context/DefaultContextSnapshot.java#L78) I think it would be more clear to get the value and either check and/or assert that it is not `null`. Otherwise, if `null` finds its way in somehow (bug?), it goes into the `setValue` method of `TLA` where it may or may not be handled well. In short, we should assert our assumption that there is no `null` value.

File: context-propagation/src/test/java/io/micrometer/context/DefaultContextSnapshotTests.java (hunk 1 above)
Comment (chemicL) [reply context-propagation/src/test/java/io/micrometer/context/DefaultContextSnapshotTests.java:73]: `should_not_fail_on_empty_thread_local` addresses point 1. I will move it closer to the other tests around null situations.
For 3 and 4 I intend to validate in a single test by calling `ContextSnapshot.setAllThreadLocals` and afterwards `ContextSnapshot.setThreadLocals` with a key.

File: context-propagation/src/main/java/io/micrometer/context/ThreadLocalAccessor.java [hunk 2]
```diff
@@ -38,13 +38,16 @@
     Object key();
 
     /**
-     * Return the current {@link ThreadLocal} value, or {@code null} if not set.
```
Comment (chemicL) [inline context-propagation/src/main/java/io/micrometer/context/ThreadLocalAccessor.java:41]: I removed the mention of `null` signifying "not set". We actually can't tell whether the value was set or not. Implementations should perform a `ThreadLocal.get()`, which upon first retrieval, will initialise the storage and set it to the initialised value - either null, or provided via the `Supplier` from `ThreadLocal.withInitial(Supplier)`.

File: context-propagation/src/main/java/io/micrometer/context/ContextSnapshot.java [hunk 3]
```diff
@@ -25,6 +26,14 @@
  * methods to propagate those values.
  *
  * <p>
+ * Implementations are disallowed to store {@code null} values. If a {@link ThreadLocal}
+ * is not set, or it's value is {@code null}, there is no way of distinguishing one from
+ * the other. In such a case, the {@link ContextSnapshot} simply must not contain a
+ * capture for the particular {@link ThreadLocal}. Implementations should filter out any
+ * {@code null} values after reading into the storage also obtained by calling
+ * {@link ContextAccessor#readValues(Object, Predicate, Map)}.
```
Comment (rstoyanchev) [inline context-propagation/src/main/java/io/micrometer/context/ContextSnapshot.java:34]:
```suggestion
 * {@link ContextAccessor#readValues(Object, Predicate, Map)}, and should likewise ignore
 * {@code null} values from {@link ContextAccessor#readValue(Object, Object)}.
```

**Review by rstoyanchev — APPROVED** [review]
Thanks for making the updates. Looks good, just one minor Javadoc update below.

**Comment by chemicL** [comment]
@rstoyanchev thank you for the comprehensive review!

**Comment by chemicL** [comment]
For future reference: The merge commit (e443fe98959bc6904c4ed4a97dc4806ea7662222) has been squashed (failure to squash+merge on my part) into c5998ba5be8f031a12fbdfa72ea65402cc300638
